Source/Parser/Functions/ArrayReduceFunction.cs

```diff
@@ -0,0 +1,83 @@
+﻿using RATools.Parser.Expressions;
+using System;
+using System.Linq;
+
+namespace RATools.Parser.Functions
+{
+    internal class ArrayReduceFunction : FunctionDefinitionExpression
+    {
+        public ArrayReduceFunction()
+            : base("array_reduce")
+        {
+            Parameters.Add(new VariableDefinitionExpression("inputs"));
+            Parameters.Add(new VariableDefinitionExpression("initial"));
+            Parameters.Add(new VariableDefinitionExpression("reducer"));
+        }
+
+        public override bool Evaluate(InterpreterScope scope, out ExpressionBase result)
+        {
+            var array = GetParameter(scope, "inputs", out result);
+            if (array == null)
+            {
+                return false;
+            }
+            if (!array.ReplaceVariables(scope, out result))
+            {
+                return false;
+            }
+            var inputs = result as IIterableExpression;
+            if (inputs == null)
+            {
+                result = new ErrorExpression("Cannot iterate over " + result.ToString(), result);
+                return false;
+            }
+
+            var accumulatorExpression = GetParameter(scope, "initial", out result);
+            if (accumulatorExpression == null)
+            {
+                return false;
+            }
+
+            var funcParam = GetFunctionParameter(scope, "reducer", out result);
+            if (funcParam == null)

```

Comment:
recommend naming this `reducer` to match the parameter name (similar to how the IterableJoiningFunction uses `predicate`.